File: desktopclient/src/main/java/com/goexp/galgame/gui/util/Tags.java
package com.goexp.galgame.gui.util;

import javafx.scene.Node;
import javafx.scene.control.Label;

import java.util.List;
import java.util.Objects;
import java.util.function.Function;
import java.util.stream.Collectors;

public class Tags {

    public static List<Node> toNodes(List<String> tag) {
        Objects.requireNonNull(tag);

        return toNodes(tag, str -> {
            var tagLabel = new Label(str);
            tagLabel.getStyleClass().add("tag");

            return tagLabel;
        });
    }

    public static List<Node> toNodes(List<String> tag, Function<String, Node> mapper) {
        Objects.requireNonNull(tag);
        Objects.requireNonNull(mapper);

        return tag.stream()
                .filter(str -> !str.isEmpty())
                .map(mapper)
                .collect(Collectors.toUnmodifiableList());
    }
}

File: desktopclient/src/main/scala/com/goexp/galgame/gui/util/Tags.scala
package com.goexp.galgame.gui.util

import javafx.scene.Node
import javafx.scene.control.Label

import java.util
import java.util.Objects
import scala.jdk.CollectionConverters._

object Tags {
  type NodeMaker = String => Node

  implicit val maker: NodeMaker = str => {
    val tagLabel = new Label(str)
    tagLabel.getStyleClass.add("tag")
    tagLabel
  }


  def toNodes(tag: String): Node = {
    Objects.requireNonNull(tag)

    toNodes(util.List.of(tag)).get(0)

  }

  def toNodes(tag: util.List[String])(implicit mapper: NodeMaker): util.List[Node] = {
    Objects.requireNonNull(tag)
    Objects.requireNonNull(mapper)
    tag.asScala.to(LazyList)
      .filter {
        _.nonEmpty
      }
      .map(mapper)
      .asJava
  }

  def toNodes(tag: List[String])(implicit mapper: NodeMaker) = {
    Objects.requireNonNull(tag)
    Objects.requireNonNull(mapper)

    tag.to(LazyList)
      .filter {
        _.nonEmpty
      }
      .map(mapper)
  }
}